Index: linux-2.6.36/drivers/usb/core/mtk/devio.c
===================================================================
--- linux-2.6.36.orig/drivers/usb/core/mtk/devio.c	2016-12-14 10:01:24.839472791 +0800
+++ linux-2.6.36/drivers/usb/core/mtk/devio.c	2016-12-14 10:12:27.335472791 +0800
@@ -928,6 +928,7 @@
 	unsigned int ep;
 	int pipe;
 	int ret;
+	struct usb_hcd *hcd = NULL;
 
 	if (get_user(ep, (unsigned int __user *)arg))
 		return -EFAULT;
@@ -937,6 +938,11 @@
 	ret = checkintf(ps, ret);
 	if (ret)
 		return ret;
+
+	hcd = bus_to_hcd(ps->dev->bus);
+	if (hcd->driver->flags & HCD_USB3)
+		return 0;
+
 	if (ep & USB_DIR_IN)
 		pipe = usb_rcvbulkpipe(ps->dev, ep & 0x7f);
 	else
